servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/DefaultNettyConnection.java

```diff
@@ -641,7 +640,7 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
                 connection.nettyChannelPublisher.channelInboundClosed();
             } else if (evt instanceof SslHandshakeCompletionEvent) {
                 connection.sslSession = extractSslSessionAndReport(ctx.pipeline(), (SslHandshakeCompletionEvent) evt,
-                        this::tryFailSubscriber, observer != null);
+                        this::tryFailSubscriber, observer != NoopConnectionObserver.INSTANCE);
```

Comment:
I was hoping `shouldReport` to go away. AFAIU, it is required because we are adding the `ConnectionObserverHandler` conditionally. Can we add the handler unconditionally?

My motivation here is to remove conditional code based on instance checks which assumes that an observer is homogeneous => an observer can not be partially NOOP.

Reply:
Because we anyway need to have some NOOP impl, I use an internal `NoopTransportObserver` as a default when no observer is configured by users. For this case we have an opportunity to optimize and skip unnecessary handlers and operators.
For the case when users add a custom observer (even partially NOOP), we don't need to optimize those use-cases. Therefore, a reference check only for out impl is enough.

Instead of adding `ConnectionObserverHandler` unconditionally, we can simplify `extractSslSessionAndReport` and look up for the handler unconditionally. But it will require us to write a test for `SecurityHandshakeObserver` every time we add a new protocol, as described in my previous comment.

Reply:
> But it will require us to write a test for SecurityHandshakeObserver every time we add a new protocol

As adding new protocols is less frequent, I think it is fair to trade-off code clarity in this case. 

> look up for the handler unconditionally.

sgtm